daemon/cluster/cluster.go

```diff
@@ -395,31 +395,32 @@ func (c *Cluster) Init(req types.InitRequest) (string, error) {
 
 	localAddr := listenHost
 
-	// If the advertise address is not one of the system's
-	// addresses, we also require a listen address.
-	listenAddrIP := net.ParseIP(listenHost)
-	if listenAddrIP != nil && listenAddrIP.IsUnspecified() {
+	// If the local address is undetermined, the advertise address
+	// will be used as local address, if it belongs to this system.
+	// If the advertise address is not local, then we need to detect
+	// a system address to use as listen and local addresses.
+	if net.ParseIP(localAddr).IsUnspecified() {
 		advertiseIP := net.ParseIP(advertiseHost)
-		if advertiseIP == nil {
-			// not an IP
-			c.Unlock()
-			return "", errMustSpecifyListenAddr
-		}
-
-		systemIPs := listSystemIPs()
 
 		found := false
-		for _, systemIP := range systemIPs {
+		for _, systemIP := range listSystemIPs() {
 			if systemIP.Equal(advertiseIP) {
+				localAddr = advertiseIP.String()
 				found = true
 				break
 			}
 		}
+
 		if !found {
-			c.Unlock()
-			return "", errMustSpecifyListenAddr
+			ip, err := c.resolveSystemAddr()
+			if err != nil {
+				c.Unlock()
+				logrus.Warnf("Could not find a local address: %v", err)
+				return "", errMustSpecifyListenAddr
+			}
+			listenHost = ip.String()
```

Comment:
In this case it is, because if the the user asked to bind to `0.0.0.0` explicitly or did not specify `--listen-addr` (in both cases `listenHost==0.0.0.0` so `true == net.ParseIP(localAddr).IsUnspecified()`), we would fail the `swarm init` (because of the adv addr locality).

If `listenHost==0.0.0.0`, and `--advertise-addr` is not local, we currently ask the user to specify a `--listen-addr`, in other words a `listenHost`. Therefore here we are sparing the user from specifying a `--listen-addr` when we can find one by ourselves.

It is true we could just take care of the local address, and my first implementation was only doing that, but I felt it was safer, more predictable to fill in what currently we ask the user to fill, which is the `--listen-addr`, and keep current logic where the local address is the listen address (when available).

So that the addresses libnetwork receive are consistent either when `--listen-addr` is specified by user or when we need user to specify the `--listen-addr` but we can do it ourselves for ease of use.